confd/pkg/backends/calico/routes.go

```diff
@@ -423,10 +449,24 @@ func (rg *routeGenerator) advertiseThisService(svc *v1.Service, ep *v1.Endpoints
 		return false
 	}
 
-	// we need to announce single IPs for services of type LoadBalancer and externalTrafficPolicy Cluster
-	if svc.Spec.Type == v1.ServiceTypeLoadBalancer && svc.Spec.ExternalTrafficPolicy == v1.ServiceExternalTrafficPolicyTypeCluster && rg.isSingleLoadBalancerIP(svc.Spec.LoadBalancerIP) {
-		logc.Debug("Advertising load balancer of type cluster because of single IP definition")
-		return true
+	// we need to announce single IPs for services of type externalTrafficPolicy Cluster.
+	// There are 2 cases inside this type:
+	// - LoadBalancer with a single IP.
+	// - Any one of externalIPs in service of type LoadBalancer or NodePort with a single IP.
+	if svc.Spec.ExternalTrafficPolicy == v1.ServiceExternalTrafficPolicyTypeCluster {
+		if svc.Spec.Type == v1.ServiceTypeLoadBalancer && rg.isSingleLoadBalancerIP(svc.Spec.LoadBalancerIP) {
+			logc.Debug("Advertising load balancer of type cluster because of single IP definition")
+			return true
+		}
+
+		if svc.Spec.Type == v1.ServiceTypeLoadBalancer || svc.Spec.Type == v1.ServiceTypeNodePort {
+			for _, extIP := range svc.Spec.ExternalIPs {
```

Comment:
If I understand correctly, this is going to advertise ALL ExternalIPs (even those which don't meet the criteria of matching a configured ExternalIP range exactly) if any of the ExternalIPs matches `isSingleExternalIP` - I think this means that any Service with one isSingleExternalIP will result in all of its ExternalIPs being advertised from every node, even if they are covered by an existing advertised route?

This is probably OK - I think it's a niche enough scenario that we shouldn't fuss too much (yet) about it, but might be worth confirming and perhaps commenting in the code that this is a known quirk of this codepath.

Reply:
I have a confusion here. Won't [this](https://github.com/projectcalico/calico/blob/21f120f1e8bcbce5877cb481dc38572e23003384/confd/pkg/backends/calico/routes.go#L240-L249) codepath chose which externalIPs to actually advertise. The one you mentioned is for if the service is needed to be exposed. Even if this comes to be true, while advertising the routes the code I pasted will see if all the externalIP are allowed by BGPConfiguration. 

Please correct me if I am wrong though. Pretty new to the project.

Reply:
@tanujd11 you're absolutely correct - it's been a while since I've looked at this part of the code! Thanks for this :)